lib/logstash/outputs/elasticsearch/http_client.rb

```diff
@@ -105,34 +138,86 @@ def close
       @pool.close
     end
 
-    private
+    
+    def calculate_property(uris, property, default, sniff_check)
+      values = uris.map(&property).uniq
 
-    #
-    # The `options` is a hash where the following symbol keys have meaning:
-    #
-    # * `:hosts` - array of String. Set a list of hosts to use for communication.
-    # * `:port` - number. set the port to use to communicate with Elasticsearch
-    # * `:user` - String. The user to use for authentication.
-    # * `:password` - String. The password to use for authentication.
-    # * `:timeout` - Float. A duration value, in seconds, after which a socket
-    #    operation or request will be aborted if not yet successfull
-    # * `:client_settings` - a hash; see below for keys.
-    #
-    # The `client_settings` key is a has that can contain other settings:
-    #
-    # * `:ssl` - Boolean. Enable or disable SSL/TLS.
-    # * `:proxy` - String. Choose a HTTP HTTProxy to use.
-    # * `:path` - String. The leading path for prefixing Elasticsearch
-    #   requests. This is sometimes used if you are proxying Elasticsearch access
-    #   through a special http path, such as using mod_rewrite.
-    def build_pool(options)
-      hosts = options[:hosts] || ["127.0.0.1"]
-      client_settings = options[:client_settings] || {}
-      timeout = options[:timeout] || 0
+      if sniff_check && values.size > 1
+        raise LogStash::ConfigurationError, "Cannot have multiple values for #{property} in hosts when sniffing is enabled!"
+      end
+
+      uri_value = values.first
+
+      default = nil if default.is_a?(String) && default.empty? # Blanks are as good as nil
+      uri_value = nil if uri_value.is_a?(String) && uri_value.empty?
+
+      if default && uri_value && (default != uri_value)
+        raise LogStash::ConfigurationError, "Explicit value for '#{property}' was declared, but it is different in one of the URLs given! Please make sure your URLs are inline with explicit values. The URLs have the property set to '#{uri_value}', but it was also set to '#{default}' explicitly"
+      end
+
+      uri_value || default
+    end
+
+    def sniffing
+      @options[:sniffing]
+    end
 
-      host_ssl_opt = client_settings[:ssl].nil? ? nil : client_settings[:ssl][:enabled]
-      urls = hosts.map {|host| host_to_url(host, host_ssl_opt, client_settings[:path], client_settings[:parameters])}
+    def user
+      calculate_property(uris, :user, @options[:user], sniffing)
+    end
+
+    def password
+      calculate_property(uris, :password, @options[:password], sniffing)
+    end
+
+    def path
+      calculated = calculate_property(uris, :path, client_settings[:path], sniffing)
```

Comment:
That would actually break in some situations.
`calculate_property` can return nils, hence the nil check. Additionally, `prepend` should be named `prepend!` because it mutates whatever is passed into it :(.

Some improvement may be possible, but I'm not so concerned with it.